README
NAME
    Perinci::Sub::GetArgs::Argv - Get subroutine arguments from command line
    arguments (@ARGV)

VERSION
    This document describes version 0.60 of Perinci::Sub::GetArgs::Argv
    (from Perl distribution Perinci-Sub-GetArgs-Argv), released on
    2014-12-19.

SYNOPSIS
     use Perinci::Sub::GetArgs::Argv;

     my $res = get_args_from_argv(argv=>\@ARGV, meta=>$meta, ...);

DESCRIPTION
    This module provides "get_args_from_argv()", which parses command line
    arguments (@ARGV) into subroutine arguments (%args). This module is used
    by Perinci::CmdLine. For explanation on how command-line options are
    processed, see Perinci::CmdLine's documentation.

FUNCTIONS
  gen_getopt_long_spec_from_meta(%args) -> [status, msg, result, meta]
    Generate Getopt::Long spec from Rinci function metadata.

    This routine will produce a "Getopt::Long" specification from Rinci
    function metadata, as well as some more data structure in the result
    metadata to help producing a command-line help/usage message.

    Function arguments will be mapped to command-line options with the same
    name, with non-alphanumeric characters changed to "-" ("-" is preferred
    over "_" because it lets user avoid pressing Shift on popular
    keyboards). For example: "file_size" becomes "file-size",
    "file_size.max" becomes "file-size-max". If function argument option
    name clashes with command-line option or another existing option, it
    will be renamed to "NAME-arg" (or "NAME-arg2" and so on). For example:
    "help" will become "help-arg" (if "common_opts" contains "help", that
    is).

    Each command-line alias ("cmdline_aliases" property) in the argument
    specification will also be added as command-line option, except if it
    clashes with an existing option, in which case this function will warn
    and skip adding the alias. For more information about "cmdline_aliases",
    see "Rinci::function".

    For arguments with type of "bool", Getopt::Long will by default also
    automatically recognize "--noNAME" or "--no-NAME" in addition to
    "--name". So this function will also check those names for clashes.

    For arguments with type array of simple scalar, "--NAME" can be
    specified more than once to append to the array.

    If "per_arg_json" setting is active, and argument's schema is not a
    "required simple scalar" (e.g. an array, or a nullable string), then
    "--NAME-json" will also be added to let users input undef (through
    "--NAME-json null") or a non-scalar value (e.g. "--NAME-json
    '[1,2,3]'"). If this name conflicts with another existing option, a
    warning will be displayed and the option will not be added.

    If "per_arg_yaml" setting is active, and argument's schema is not a
    "required simple scalar" (e.g. an array, or a nullable string), then
    "--NAME-yaml" will also be added to let users input undef (through
    "--NAME-yaml '~'") or a non-scalar value (e.g. "--NAME-yaml '[foo,
    bar]'"). If this name conflicts with another existing option, a warning
    will be displayed and the option will not be added. YAML can express a
    larger set of values, e.g. binary data, circular references, etc.

    Will produce a hash (Getopt::Long spec), with "func.specmeta",
    "func.opts", "func.common_opts", "func.func_opts" that contain extra
    information ("func.specmeta" is a hash of getopt spec name and a hash of
    extra information while "func.*opts" lists all used option names).

    Arguments ('*' denotes required arguments):

    *   args => *hash*

        Reference to hash which will store the result.

    *   common_opts => *hash*

        Common options.

        A hash where the values are hashes containing these keys: "getopt"
        (Getopt::Long option specification), "handler" (Getopt::Long
        handler). Will be passed to "get_args_from_argv()". Example:

         {
             help => {
                 getopt  => 'help|h|?',
                 handler => sub { ... },
                 summary => 'Display help and exit',
             },
             version => {
                 getopt  => 'version|v',
                 handler => sub { ... },
                 summary => 'Display version and exit',
             },
         }

    *   ignore_converted_code => *bool* (default: 0)

        Whether to ignore coderefs converted to string.

        Across network through JSON encoding, coderef in metadata (e.g. in
        "cmdline_aliases" property) usually gets converted to string "CODE".
        In some cases, like for tab completion, this is pretty harmless so
        you can turn this option on. For example, in the case of
        "cmdline_aliases", the effect is just that command-line aliases code
        are not getting executed, but this is usually okay.

    *   meta* => *hash*

        Rinci function metadata.

    *   meta_is_normalized => *bool*

    *   per_arg_json => *bool* (default: 0)

        Whether to add --NAME-json for non-simple arguments.

        Will also interpret command-line arguments as JSON if assigned to
        function arguments, if arguments' schema is not simple scalar.

    *   per_arg_yaml => *bool* (default: 0)

        Whether to add --NAME-yaml for non-simple arguments.

        Will also interpret command-line arguments as YAML if assigned to
        function arguments, if arguments' schema is not simple scalar.

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

     (any)

  get_args_from_argv(%args) -> [status, msg, result, meta]
    Get subroutine arguments (%args) from command-line arguments (@ARGV).

    Using information in Rinci function metadata's "args" property, parse
    command line arguments @argv into hash %args, suitable for passing into
    subroutines.

    Currently uses Getopt::Long's GetOptions to do the parsing.

    As with GetOptions, this function modifies its "argv" argument, so you
    might want to copy the original "argv" first (or pass a copy instead) if
    you want to preserve the original.

    See also: gen_getopt_long_spec_from_meta() which is the routine that
    generates the specification.

    Arguments ('*' denotes required arguments):

    *   allow_extra_elems => *bool* (default: 0)

        Allow extra/unassigned elements in argv.

        If set to 1, then if there are array elements unassigned to one of
        the arguments, instead of generating an error, this function will
        just ignore them.

        This option will be passed to Perinci::Sub::GetArgs::Array's
        allow_extra_elems.

    *   args => *hash*

        Specify input args, with some arguments preset.

    *   argv => *array*

        If not specified, defaults to @ARGV

    *   common_opts => *hash*

        Common options.

        A hash where the values are hashes containing these keys: "getopt"
        (Getopt::Long option specification), "handler" (Getopt::Long
        handler). Will be passed to "get_args_from_argv()". Example:

         {
             help => {
                 getopt  => 'help|h|?',
                 handler => sub { ... },
                 summary => 'Display help and exit',
             },
             version => {
                 getopt  => 'version|v',
                 handler => sub { ... },
                 summary => 'Display version and exit',
             },
         }

    *   ignore_converted_code => *bool* (default: 0)

        Whether to ignore coderefs converted to string.

        Across network through JSON encoding, coderef in metadata (e.g. in
        "cmdline_aliases" property) usually gets converted to string "CODE".
        In some cases, like for tab completion, this is harmless so you can
        turn this option on.

    *   meta* => *hash*

    *   meta_is_normalized => *bool* (default: 0)

        Can be set to 1 if your metadata is normalized, to avoid duplicate
        effort.

    *   on_missing_required_args => *code*

        Execute code when there is missing required args.

        This can be used to give a chance to supply argument value from
        other sources if not specified by command-line options.
        Perinci::CmdLine, for example, uses this hook to supply value from
        STDIN or file contents (if argument has "cmdline_src" specification
        key set).

        This hook will be called for each missing argument. It will be
        supplied hash arguments: (arg => $the_missing_argument_name, args =>
        $the_resulting_args_so_far, spec => $the_arg_spec).

        The hook can return true if it succeeds in making the missing
        situation resolved. In this case, this function will not report the
        argument as missing.

    *   per_arg_json => *bool* (default: 0)

        Whether to recognize --ARGNAME-json.

        This is useful for example if you want to specify a value which is
        not expressible from the command-line, like 'undef'.

         % script.pl --name-json 'null'

        But every other string will need to be quoted:

         % script.pl --name-json '"foo"'

        See also: per_arg_yaml. You should enable just one instead of
        turning on both.

    *   per_arg_yaml => *bool* (default: 0)

        Whether to recognize --ARGNAME-yaml.

        This is useful for example if you want to specify a value which is
        not expressible from the command-line, like 'undef'.

         % script.pl --name-yaml '~'

        See also: per_arg_json. You should enable just one instead of
        turning on both.

    *   strict => *bool* (default: 1)

        Strict mode.

        If set to 0, will still return parsed argv even if there are parsing
        errors (reported by Getopt::Long). If set to 1 (the default), will
        die upon error.

        Normally you would want to use strict mode, for more error checking.
        Setting off strict is used by, for example, Perinci::Sub::Complete
        during completion where the command-line might still be incomplete.

        Should probably be named "ignore_errors". :-)

    Return value:

    Returns an enveloped result (an array).

    First element (status) is an integer containing HTTP status code (200
    means OK, 4xx caller error, 5xx function error). Second element (msg) is
    a string containing error message, or 'OK' if status is 200. Third
    element (result) is optional, the actual result. Fourth element (meta)
    is called result metadata and is optional, a hash that contains extra
    information.

     (any)

    Error codes:

    *   400 - Error in Getopt::Long option specification, e.g. in
        common_opts.

    *   500 - failure in GetOptions, meaning argv is not valid according to
        metadata specification (only if 'strict' mode is enabled).

    *   501 - coderef in cmdline_aliases got converted into a string,
        probably because the metadata was transported (e.g. through
        Riap::HTTP/Riap::Simple).

FAQ
TODO
    Option to enable json/yaml for nullable simple scalar (to enable
    "--str-json '~'").

SEE ALSO
    Perinci

HOMEPAGE
    Please visit the project's homepage at
    <https://metacpan.org/release/Perinci-Sub-GetArgs-Argv>.

SOURCE
    Source repository is at
    <https://github.com/perlancar/perl-Perinci-Sub-GetArgs-Argv>.

BUGS
    Please report any bugs or feature requests on the bugtracker website
    <https://rt.cpan.org/Public/Dist/Display.html?Name=Perinci-Sub-GetArgs-A
    rgv>

    When submitting a bug or request, please include a test-file or a patch
    to an existing test-file that illustrates the bug or desired feature.

AUTHOR
    perlancar <perlancar@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by perlancar@cpan.org.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perinci-Sub-GetArgs-Agv

0.61    2015-01-03 (PERLANCAR)

	- No functional changes.

	- Rebuild (Generate TODO.md).


0.60    2014-12-19 (PERLANCAR)

        - Shut up warning from Getopt::Long (by locally installing a
          $SIG{__WARN__} handler) when strict=0.


0.59    2014-11-20 (PERLANCAR)

	- Add common_opts/neg_opts/pos_opts in func.specmeta, to help
          Perinci::Sub::To::CLIOptSpec.


0.58    2014-11-19 (PERLANCAR)

	- Add --isnt-FOO and --arent-FOO as negation forms for --is-FOO and
	  --are-FOO (and vice versa) (currently untested).


0.57    2014-10-24 (PERLANCAR)

	- Add --ARG-base64 for buf arguments.

	- [ux] Change negated options for with-foo -> without-foo (and vice
	  versa) instead of the default nowith-foo + no-with-foo.


0.56    2014-10-23 (PERLANCAR)

	- No functional changes.

	- Typo: unimplemented status is 501, not 502.


0.55    2014-10-11 (PERLANCAR)

	- Support is_flag cmdline_alias property.

	- [Bugfix] cmdline alias handler should also get the same treatment
	  as arg handler.


0.54    2014-10-10 (PERLANCAR)

	- Add support for element_meta (Rinci 1.1.56).


0.53    2014-10-10 (PERLANCAR)

	- No functional changes.

        - Optimize gen_getopt_long_spec_from_meta() by avoiding a bunch of calls
          to parse_getopt_long_opt_spec(), making it about 60% faster for a
          sample metadata of 5 arguments + 1 cmdline_alias.

        - Add a couple of tests.


0.52    2014-10-09 (PERLANCAR)

	- Support argument submetadata (Rinci 1.1.55).


0.51     2014-09-03 (PERLANCAR)

         - Add option: ignore_converted_code (this option is needed e.g. by
           Perinci::Sub::Complete to keep tab completion working even when there
           is coderef cmdline_aliases property in arg spec).


0.50     2014-08-06 (SHARYANTO)

         - Revert previous feature (accept _genres).


0.49     2014-08-06 (SHARYANTO)

         - Remove orig_spec key, because we now no longer re-order opts in ospec.

         - get_args_from_argv(): Accept _genres (a temporary dirty hack)
           argument to avoid doing gen_getopt_long_spec_from_meta() again.


0.48     2014-07-25 (SHARYANTO)

         [ENHANCEMENTS]

         - For arguments of type 'array of simple scalars', generate
           Getopt::Long option spec as 'NAME=s@' instead of just 'NAME=s'; this
           makes it more evident that the option is repeatable.


0.47     2014-07-23 (SHARYANTO)

         - No functional changes.

         - Adjust to Getopt::Long::Util 0.77.


0.46     2014-07-23 (SHARYANTO)

         [INCOMPATIBLE CHANGES]

         - Adjust to Getopt::Long::Util 0.46 (option ordering).

         - Change 'common_opts' argument form so it is now the same as
           Perinci::CmdLine's.


0.45     2014-07-22 (SHARYANTO)

         - Don't attempt parsing argv as json/yaml if array of simple scalar.


0.44     2014-07-18 (SHARYANTO)

         - These changes are to aid completion.

         - gen_getopt_long_spec_from_meta(): Provide 'func.opts_by_arg',
           'func.opts_by_common', 'func.common_opts' & 'func.func_opts'.

         - get_args_from_argv(): Return gen_getopt_long_spec_from_meta result in
           func.gen_getopt_long_spec_result to avoid caller having to call
           gen_getopt_long_spec_from_meta() separately.

         [INCOMPATIBLE CHANGES]

         - gen_getopt_long_spec_from_meta(): func.*opts are now prefixed with
           --/-.


0.43     2014-07-18 (SHARYANTO)

         - gen_getopt_long_spec_from_meta(): Add 'parsed' key (result from
           Getopt::Long::Util's parse_getopt_long_opt_spec()). Add 'alias_for'
           key for alias options to trace back original option. Both to aid
           help/usage generation.

         - [Bugfix] Some bug fixes.


0.42     2014-07-18 (SHARYANTO)

         - gen_getopt_long_spec_from_meta(): Add 'orig_spec' to common_opts'
           alias specmeta to be able to trace back the original key (to aid
           help/usage generation).


0.41     2014-07-18 (SHARYANTO)

         - gen_getopt_long_spec_from_meta(): Add is_alias to alias specmeta, add
           {,non}code_aliases to non-alias specmeta (to aid help/usage
           generation).


0.40     2014-07-18 (SHARYANTO)

         - No functional changes.

         - Revert previous decision: Deliberately not including YAML::Syck &
           JSON as prereq to reduce deps. It complicates testing [CT].


0.39     2014-07-17 (SHARYANTO)

         - To reduce dependencies, deliberately no longer lists JSON and
           YAML::Syck as dependencies. These modules are only required when
           'per_arg_json' and 'per_arg_yaml' are enabled.

         [BUG FIXES]

         - Only try to parse positional command-line arguments as JSON/YAML when
           per_arg_yaml/per_arg_json is true.


0.38     2014-07-12 (SHARYANTO)

         - Add option: 'args', to set initial value for the resulting arguments.


0.37     2014-07-09 (SHARYANTO)

         [INCOMPATIBLE CHANGES]

         - Changes the interface and the parsing rules slightly. Since I want to
           remove --action and the subtle difference of 'extra_getopts_before'
           and 'extra_getopts_after'. Now we avoid clashes between common
           options like '--help' and function arguments of the same name (they
           will be renamed to '--help-arg', or '--help-arg2' and so on upon
           clash). Clash will also be avoided between '--ARGNAME' and
           '--ARGNAME-json' should that happen.

         [REMOVED FEATURES]

         - Remove 'check_required_args' option. Now function still returns 200
           on missing args. Missing args are already reported in the result
           metadata ('func.missing_args').

         - Remove 'extra_getopts_before' and 'extra_getopts_after', replace it
           with 'common_opts' (a hash of Getopt::Long specification name and
           handlers).

         [ENHANCEMENTS]


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Get subroutine arguments from command line arguments (\@ARGV)",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Perinci-Sub-GetArgs-Argv",
  "dist_version" => "0.23",
  "license" => "perl",
  "module_name" => "Perinci::Sub::GetArgs::Argv",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Data::Sah" => "0.02",
    "JSON" => 0,
    "Log::Any" => 0,
    "Perinci::Sub::GetArgs::Array" => "0.09",
    "Perinci::Sub::Util" => 0,
    "Rinci" => "v1.1.10",
    "YAML::Syck" => 0,
    "perl" => "5.010001"
  },
  "script_files" => [],
  "test_requires" => {
    "Data::Clone" => 0,
    "Test::More" => "0.98"
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.029.
use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Get subroutine arguments from command line arguments (\@ARGV)",
  "AUTHOR" => "perlancar <perlancar\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "Perl::osnames" => "0.09"
  },
  "DISTNAME" => "Perinci-Sub-GetArgs-Argv",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010001",
  "NAME" => "Perinci::Sub::GetArgs::Argv",
  "PREREQ_PM" => {
    "Data::Clean::FromJSON" => "0.11",
    "Data::Sah::Normalize" => 0,
    "Function::Fallback::CoreOrPP" => 0,
    "Getopt::Long::Util" => "0.77",
    "JSON" => 0,
    "Perinci::Sub::GetArgs::Array" => "0.14",
    "Perinci::Sub::Normalize" => "0.07",
    "Perinci::Sub::Util" => 0,
    "Rinci" => "1.1.57",
    "YAML::Syck" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Clean::JSON" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "0.98"
  },
  "VERSION" => "0.61",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Clean::FromJSON" => "0.11",
  "Data::Clean::JSON" => 0,
  "Data::Sah::Normalize" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Function::Fallback::CoreOrPP" => 0,
  "Getopt::Long::Util" => "0.77",
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON" => 0,
  "Perinci::Sub::GetArgs::Array" => "0.14",
  "Perinci::Sub::Normalize" => "0.07",
  "Perinci::Sub::Util" => 0,
  "Perl::osnames" => "0.09",
  "Rinci" => "1.1.57",
  "Test::More" => "0.98",
  "YAML::Syck" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



